pkg/ebpf/tracer_windows.go

```diff
@@ -179,6 +218,82 @@ func (t *Tracer) GetStats() (map[string]interface{}, error) {
 	}, nil
 }
 
+// NewDDAPIFilter returns a filter we can apply to the driver
+func newDDAPIFilter(direction C.uint64_t, ifaceIndex int, isIPV4 bool) (fd C.struct__filterDefinition) {
+	fd.filterVersion = C.DD_FILTER_SIGNATURE
+	fd.size = C.sizeof_struct__filterDefinition
+	fd.direction = direction
+
+	if isIPV4 {
+		fd.af = windows.AF_INET
+		fd.v4InterfaceIndex = (C.ulonglong)(ifaceIndex)
+	} else {
+		fd.af = windows.AF_INET6
+		fd.v6InterfaceIndex = (C.ulonglong)(ifaceIndex)
+	}
+
+	return fd
+}
+
+func (t *Tracer) setFilter(fd C.struct__filterDefinition) (err error) {
+	var id int64
+	err = windows.DeviceIoControl(t.driverHandle, C.DDFILTER_IOCTL_SET_FILTER, (*byte)(unsafe.Pointer(&fd)), uint32(unsafe.Sizeof(fd)), (*byte)(unsafe.Pointer(&id)), uint32(unsafe.Sizeof(id)), nil, nil)
+	if err != nil {
+		return log.Errorf("Failed to set filter: %s\n", err.Error())
+	}
+	return
+}
+
+// To capture all traffic for an interface, we create an inbound/outbound traffic filter
+// for both IPV4 and IPV6 traffic going to that interface
+func createFiltersForInterface(iface net.Interface) (filters []C.struct__filterDefinition) {
+	filters = append(filters, newDDAPIFilter(C.DIRECTION_INBOUND, iface.Index, true))
+	filters = append(filters, newDDAPIFilter(C.DIRECTION_OUTBOUND, iface.Index, true))
+	filters = append(filters, newDDAPIFilter(C.DIRECTION_INBOUND, iface.Index, false))
+	filters = append(filters, newDDAPIFilter(C.DIRECTION_OUTBOUND, iface.Index, false))
+	return
+}
+
+func (t *Tracer) prepareDriverFilters() (err error) {
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		log.Errorf("Error getting interfaces: %s\n", err.Error())
+		return
+	}
+
+	for _, i := range ifaces {
+		log.Debugf("Setting filters for interface: %s [%+v]", i.Name, i)
+
+		for _, filter := range createFiltersForInterface(i) {
+			err = t.setFilter(filter)
+			if err != nil {
+				log.Warnf("Failed to set filter [%+v] on interface [%+v]: %s\n", filter, i, err.Error())
+			}
+		}
+	}
+	return nil
+}
+
+// Add buffers to Tracer object. Even though the windows API will actually
+// keep the buffers, add it to the tracer so that golang doesn't garbage collect
+// the buffers out from under us
+func (t *Tracer) prepareReadBuffers() (err error) {
+	for i := 0; i < totalReadBuffers; i++ {
+		err = windows.ReadFile(t.driverHandle, t.bufs[i].data[:], nil, &t.bufs[i].ol)
+		if err != nil {
+			if err != windows.ERROR_IO_PENDING {
```

Comment:
What does it mean that if the err was `ERROR_IO_PENDING`?  The buffer will eventually be created?  We need to try again?

Reply:
`ERROR_IO_PENDING` is the windows error code indicating that an asynchronous IO operation has been initiated.
Think of it as `EWOULDBLOCK` on steroids.  The call to `ReadFile()` completes immediately.  When there is data available, the caller will be signaled using the notification method that the caller chose (in this case a completion port will be signalled).

Be happy to go into more detail if you want, but probably not here ;)